azure/Guidelines.md

```diff
@@ -0,0 +1,862 @@
+
+Cloud applications embrace failure. Therefore, to enable customers to write fault-tolerant applications, <b>all</b> service operations (including POST) <b>must</b> be idempotent. Implementing services in an idempotent manner, with an "exactly once" semantic, enables developers to retry requests without the risk of unintended consequences.
+
+#### Exactly Once Behavior = Client Retries & Service Idempotency
+> :white_check_mark: **DO** ensure that __all__ HTTP methods are idempotent.
+ 
+> :ballot_box_with_check: **YOU SHOULD** use PUT or PATCH to create a resource as these HTTP methods are easy to implement, allow the customer to name their own resource, and are idempotent.
+ 
+> :heavy_check_mark: **YOU MAY** use POST to create a resource but you must make it idempotent and, of course, the response __must__ return the URL of the created resource with a 201-Created. One way to make POST idempotent is to use the Repeatability-Request-ID & Repeatability-First-Sent headers (See [Repeatability of requests](#Repeatability-of-requests)).
+
+##### HTTP Return Codes
+> :white_check_mark: **DO** adhere to the return codes in the following table when the method completes synchronously and is successful:
+
+Method | Description | Response Status Code 
+-------|-------------|---------------------
+PATCH  | Create/Modify the resource with JSON Merge Patch | 200-OK, 201-Created
+PUT    | Create/Replace the *whole* resource | 200-OK, 201-Created 
+POST   | Create new resource (ID set by service) | 201-Created with URL of created resource
+GET    | Read (i.e. list) a resource collection | 200-OK
+GET    | Read the resource | 200-OK 
+DELETE | Remove the resource | 204-No Content\; avoid 404-Not Found 
+
+> :white_check_mark: **DO** return status code ```202-Accepted``` and follow the guidance in [Long Running Operations & Jobs](#long-running-operations--jobs) when a PUT, POST, or DELETE method completes asynchronously
+
+> :white_check_mark: **DO** treat method names as case sensitive and should always be in uppercase
+> 
+> :white_check_mark: **DO** return the state of the resource after a PUT, PATCH, POST, or GET operation with a ```200-OK``` or ```201-Created```.
+> 
+> :white_check_mark: **DO** return a ```204-No Content``` without a resource/body for a DELETE operation (even if the URL identifies a resource that does not exist; do not return ```404-Not Found```)
+> 
+> :white_check_mark: **DO** support caching and optimistic concurrency by honoring the the if-match, if-none-match, if-modified-since, and if-unmodified-since request headers and by returning the etag and last-modified response headers
+
```

Comment:
Consider adding:

“DO return a 404 when the user does not have access to the resource.  Returning a 404, vs. a 403, prevents a malicious request from learning anything about resources based on error codes.”

Reply:
Added.

Reply:
the safely of this additional recommendation isn't clear to me. Doesn't returning 404 on "right resource, no access" similarly enable a fuzzer to learn something?" isn't 403 beneficial here too?

Reply:
I believe ARM returns a 403 in this case at present across all RPs in the management plane if the caller does not have permissions to a scope regardless of whether that scope exists or not.  I had to investigate a customer incident just last week where this behavior was observed.

Reply:
Perhaps the wording was unclear, but the intent is to return a 404 _whether or not the resource exists_ if the caller does not have the appropriate permissions.  In other words, check permission first, but if permission fails that return 404.

Reply:
understood. I recommend we follow the ARM convention as per Michael. The Azure Notebooks Resource implementation (my team's stuff) follows the always return 403 guidance.

Reply:
I think in the data plane the 404 for permission denied is more appropriate, even if it is different than the ARM guidance.  In ARM it makes sense that the tooling should differentiate because that information is likely useful and/or important for tooling or automation built on the ARM APIs.  For the data plane we have *many* more use cases and in many of them it makes sense to not reveal information in many of those cases.

Reply:
to expand - the resource to which you don't have permission in the data plane is likely to be content that belongs to a tenant organization, whereas in the control plane it is just a cloud resource.  Revealing the existence (or not) of particular pieces of content/data is more sensitive in those cases.

Reply:
I think there is some confusion here -- ARM does not return 403s only when a resource exists but the user is not authorized.  It always returns 403s when a user doesn't have access to the scope (regardless of whether or not that scope exists or not).  The key requirement here is to not disclose additional details to an unauthorized user -- being consistent accomplishes that regardless of whether or not you consistently return a 403 or a 404.  Returning a 403 is consistent with current ARM semantics which RPs already need to account for both in a producing and consumption role.

Reply:
We updated the guidance to be in alignment with ARM. 
Resolving this thread.